consensus/src/txn_notifier.rs

```diff
@@ -48,28 +49,40 @@ impl MempoolNotifier {
 impl TxnNotifier for MempoolNotifier {
     async fn notify_failed_txn(
         &self,
-        txns: Vec<SignedTransaction>,
+        user_txns: Vec<SignedTransaction>,
         compute_results: &StateComputeResult,
+        block_gas_limit_enabled: bool,
     ) -> Result<(), MempoolError> {
         let mut rejected_txns = vec![];
 
-        if txns.is_empty() {
+        if user_txns.is_empty() {
             return Ok(());
         }
         let compute_status = compute_results.compute_status();
-        if txns.len() + 2 != compute_status.len() {
+        let expected_len = if compute_results.has_reconfiguration() && block_gas_limit_enabled {
```

Comment:
Why this needs to be special cased for `block_gas_limit_enabled` i.e., when block gas limit is not enabled, why we have the state checkpoint txn in case of epoch change and not in case of block gas limit is enabled?

Reply:
It is due to compatibility reasons. 
Before block gas limit, the state checkpoint is added before execution for any block. If a block (metadata txn) triggers reconfiguration, its checkpoint is ignored and has status as retry (same as all user txns).
After block gas limit, since we need to cut the block and have state checkpoint in the end, the checkpoint in inserted after execution. In the case of reconfiguration, we do not insert it since it is useless anyway. Otherwise, it is inserted at the end of the block (excluding txns that have status retry).